### ISSUE TITLE
Issue with sources/modifiers

### ISSUE THREAD

**Issue by RobertVejvoda** [dotnet/templating#398]
Having the following folder structure as my template:

- Root
  - Project1
      - file1.txt 
      - file2.txt
  - Project2

I want to exclude a file called *file1.txt* from Project1. Trying to define this in my template.json file:
`
    "sources": [
    {
      "modifiers": [ { "condition": "(1==1)", "exclude": "./Project1/file1.txt" } ]
     }
     ]
`
After registering my template with `dotnet new -i templ` and processing it `dotnet new templ -n MySolution` the file is in the generated directory.

**Comment by mlorbetske** [comment]
@seancpeters can you take a look please?

**Comment by seancpeters** [comment]
Hi @RobertVejvoda,
If you slightly modify your sources config to have the exclude value be an array, this will work as expected:
```
"sources": [ 
  { 
    "modifiers": [ 
      { 
        "condition": "(1==1)", 
        "exclude": [ "./Project1/file1.txt" ]
      } 
    ]
  } 
]
```
Also, the condition is optional, so you can exclude the condition since you always want it to be true:
```
"sources": [ 
  { 
    "modifiers": [ 
      { 
        "exclude": [ "./Project1/file1.txt" ]
      } 
    ]
  } 
]
```

**Comment by seancpeters** [comment]
I need to clarify: In the future it will be ok to have the config as in your original post. Looking into this I discovered a bug in how the excludes (and includes and copyonly patterns are processed). What is supposed to happen is that if the exclude value is a single string, we check if that string is the name of a file in the same directory as the template.json file. 
If that file exists, the contents of that file are read, assuming each line in the file is an exclude pattern.
If the file doesn't exist, then the string itself is treated as the exclude pattern.

But if the exclude in template.json is an array, each array entry is assumed to be an exclude pattern.

I'll be checking in a bug fix for this soon.

**Comment by RobertVejvoda** [comment]
Hi @seancpeters, 
thanks for your explanation. Unfortunately the exclude array pattern isn't working either (for me).  Please consider this:
```
"sources": [
    {
      "source": "./",
      "include": ["**/*"],
      "exclude": [ "**/[Bb]in/", "**/[Oo]bj/", "**/.template.config/**/*", "**/*.filelist", "**/*.config" ],
      "modifiers": [
        {
          "condition": "(Framework==\"core\")",
          "include": "./TemplateName.Web/app.config"
        },
        {
          "condition": "(Framework==\"full\")",
          "include": "./TemplateName.Web/web.config"
        }
      ]
    }
  ],
"symbols": {
    "Framework": {
      "type": "parameter",
      "datatype": "choice",
      "choices": [
        {
          "choice": "core",
          "description": "Target netstandard1.6"
        },
        {
          "choice": "full",
          "description": "Target net461"
        }
      ],
      "defaultValue": "core"
    }
}
```
So I don't want not include .config files at all first, and then only add the one appropriate for the target framework. Another problem here is that it also copies bin, obj and other folders in the target directory even though they're excluded. 

If you need I can send you the full zip file together with template.json.
Thanks for your help!
Rob

**Comment by seancpeters** [comment]
Hi @RobertVejvoda,
Regarding the bin, obj, & other folders being copied, that can be resolved by removing your top-level include & exclude directives, i.e.:
```
      "include": ["**/*"],
      "exclude": [ "**/[Bb]in/", "**/[Oo]bj/", "**/.template.config/**/*", "**/*.filelist", "**/*.config" ],
```
We have defaults setup that are effectively what you're trying to accomplish, but your includes override the default includes, and your excludes override the default excludes. Your includes are exactly the same as the defaults, so it won't hurt to keep that line, but it's not strictly necessary. 

Your excludes are almost the same as the defaults, the defaults being:
```
            "**/[Bb]in/**",
            "**/[Oo]bj/**",
            "**/.template.config/**",
            "**/*.filelist",
            "**/*.user",
            "**/*.lock.json"
```
The main differences are:
- Needing trailing ** on the bin & obj excludes.
- Your final exclude of "**/*.config"

It's probably easiest to use the default excludes, and then conditionally exclude the specific config that is unwanted due to the framework version. The final sources & symbols sections of the config would look like this:
```
  "sources": [
    {
      "modifiers": [
        {
          "condition": "(Framework==\"core\")",
          "exclude": ["TemplateName.Web/web.config"]
        },
        {
          "condition": "(Framework==\"full\")",
          "exclude": ["TemplateName.Web/app.config"]
        }
      ]
    }
  ],
  "symbols": {
    "Framework": {
      "type": "parameter",
      "datatype": "choice",
      "choices": [
        {
          "choice": "core",
          "description": "Target netstandard1.6"
        },
        {
          "choice": "full",
          "description": "Target net461"
        }
      ],
      "defaultValue": "core"
    }
  }
```
Note: I made the excludes be arrays to cause the literal filenames to be excluded, and avoid them trying to be interpreted as files which contain the exclude lists.

If this still isn't working as desired, please let me know - perhaps then it'll be time to get a zip from you, as I may be missing some of the context of your template.

**Comment by seancpeters** [comment]
Hi @RobertVejvoda,
I wanted to follow up a bit more on this issue. We discovered a bug based on your original template setup which caused the include modifiers to not override the source excludes. That will be fixed as of pull request #413 (not yet merged).

Using the code in that fix, your template.json could be more like you had originally posted. I've tested this sources section of the config against the PR, with a trivial template:
```
  "sources": [
    {
      "source": "./",
      "include": ["**/*"],
      "exclude": [ 
        "**/[Bb]in/**",
        "**/[Oo]bj/**",
        "**/.template.config/**",
        "**/*.filelist",            
        "**/*.user",
        "**/*.lock.json",
        "**/*.config"
      ],
      "modifiers": [
        {
          "condition": "(Framework==\"core\")",
          "include": ["TemplateName.Web/app.config"]
        },
        {
          "condition": "(Framework==\"full\")",
          "include": ["TemplateName.Web/web.config"]
        }
      ]
    }
  ],
```

Note that I setup the source excludes to be our internal defaults, plus the extra one for excluding configs, i.e.: `"**/*.config"`
Again, this won't work without the code in the PR I mentioned above, but should offer more flexibility in template authoring.

**Comment by mlorbetske** [comment]
Closing this here since #413 has been merged
